### PR TITLE
Add flush flag & leak detector

### PR DESCRIPTION
**Pull request by e5l**
- Add flush return flag to indicate if we can reuse the buffer
- introduce leak detector

### CODE REVIEW

File: core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+import kotlinx.io.buffer.*
+import kotlinx.io.pool.*
+
+/**
+ * The pool that tracks every borrowed object and check if it belongs to this pool.
+ *
+ * It also verifies if this pool is empty at close.
+ */
+internal class LeakDetectingPool(private val size: Int = DEFAULT_BUFFER_SIZE) : ObjectPool<Buffer> {
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt]: Looks like it should belong to test module

File: core/commonTest/src/kotlinx/io/InputPrimitivesTest.kt [hunk 2]
```diff
@@ -14,13 +14,15 @@ class InputPrimitivesTest {
         limit: Int, seed: Long = 0L, body: Input.() -> Unit
     ) = prefetchSizes.forEach { prefetchSize ->
         bufferSizes.forEach { size ->
+            val input = sequentialLimitedInput(fetchSizeLimit, pool, limit, seed)
```
Comment (qwwdfsad) [inline core/commonTest/src/kotlinx/io/InputPrimitivesTest.kt]: `use`?

File: core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt (hunk 1 above)
Comment (e5l) [reply core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt]: Yep, but we can't share them across test modules

File: core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+    override fun close() {
+        val liveBuffers = allocated.size - removed.size
+        check(liveBuffers == 0) {
+            "Memory leak: $liveBuffers buffers still in use. Total allocated: ${allocated.size}"
+        }
+    }
+
+    companion object {
+        private var poolCounter: Byte = 0
```
Comment (qwwdfsad) [inline core/commonMain/src/kotlinx/io/internal/LeakDetectingPool.kt]: unused

**Comment by e5l** [comment]
closed as obsolete

**Pull request closed without merging**